[SYCL] Add arithmetic overloaded operators for annotated_arg (#11117)
intel/llvm: sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_arg.asciidoc

    // instances of propertyT
    template<typename propertyT>
    static constexpr /*unspecified*/ get_property();

  template <typename T2, typename PropertyList2>

>   auto operator+(const annotated_arg<T2, PropertyList2> &other) const;

Reviewer: This isn't useful when I want to add an annotated_arg<T> to a T.  See my comment in Greg's chain for a solution to that.

template <typename T, typename PropertiesA, typename PropertiesB>
annotated_arg(annotated_arg<T, PropertiesA>, PropertiesB>) ->
    annotated_arg<T, /* a type that combines the properties of PropertiesA and PropertiesB */>;

// Overloaded operators for raw type
template<typename T, typename PropertyList, typename T2, typename R = decltype(std::declval<T>() + std::declval<T2>())>
R operator+(const annotated_arg<T, PropertyList> &a, const T2 b);

template<typename T, typename PropertyList, typename T2, typename R = decltype(std::declval<T>() - std::declval<T2>())>
R operator-(const annotated_arg<T, PropertyList> &a, const T2 b);

template<typename T, typename PropertyList, typename T2, typename R = decltype(std::declval<T>() * std::declval<T2>())>
R operator*(const annotated_arg<T, PropertyList> &a, const T2 b);

template<typename T, typename PropertyList, typename T2, typename R = decltype(std::declval<T>() / std::declval<T2>())>

> R operator/(const annotated_arg<T, PropertyList> &a, const T2 b);

Author: I tested this case and it seems like the `annotated_arg<T>` is implicitly converted to `T2` and then the overloaded operator for `T2` is used. So there is no need to add an overloaded operator for this case.

> R operator/(const annotated_arg<T, PropertyList> &a, const T2 b);

Reviewer 2: It seems very surprising to me that you can rely on implicit conversion for:

```
ac_int<int> ac1{1}, ac2{2};
annotated_arg<ac_int<int>> ann_a1{ac1}, ann_a2{ac2};
ac_int<int> ac3 = ac1 + ann_a2;
```

while at the same time implicit conversion doesn't work for:

```
ac_int<int> ac1{1}, ac2{2};
annotated_arg<ac_int<int>> ann_a1{ac1}, ann_a2{ac2};
ac_int<int> ac3 = ann_a1 + ac2;
```

Are you sure about this?  If so, what causes the assymetry?

> R operator/(const annotated_arg<T, PropertyList> &a, const T2 b);

Author: I tested this and the assymetry is present anytime a class is templated on a user-defined class. For this code: 
```c++
struct Custom {
    int value;
    Custom(int value) : value(value){};
    Custom operator+(const Custom &rhs) const {
        return Custom(value + rhs.value);
    }
};

template<typename T>
class CustomClass {
    T value;
    CustomClass(T val) : value(val) {}
    operator T() {
        return value;
    }
};

int main() {
    CustomClass<Custom> myInt(Custom(42));
    Custom result = myInt + Custom(8);
    assert(result.value == 50 && "result value does not match");
    return 0;
}
```
This produces the following error: 
```
temp.cpp:31:27: error: invalid operands to binary expression ('CustomClass<Custom>' and 'Custom')
   31 |     Custom result = myInt + Custom(8);
      |                     ~~~~~ ^ ~~~~~~~~~
```
If the line is instead: `Custom result = Custom(8) + myInt`, there is no error. Also, if the templated parameter for CustomClass is int rather a user-defined class, the above error is not produced. I am not sure what is the cause for this assymetry when a user-defined class is used as the template parameter.

> R operator/(const annotated_arg<T, PropertyList> &a, const T2 b);

Reviewer 2: Do you want to handle a case like this?

```
#include <utility>

template<typename T>
struct Custom {
    T value;
    Custom(T value) : value(value){};
};

template<typename T>
Custom<T> operator+(const Custom<T> &lhs, const Custom<T> &rhs) {
    return Custom<T>(lhs.value + rhs.value);
}

template<typename T>
struct annotated_arg {
    T value;
    annotated_arg(T val) : value(val) {}
    operator T() {
        return value;
    }
};

template<typename T, typename T2,
         typename R = decltype(std::declval<T>() + std::declval<T2>())>
R operator+(const annotated_arg<T> &a, const T2 &b) {
    return a.value + b;
}

#if 0
template<typename T, typename T2,
         typename R = decltype(std::declval<T>() + std::declval<T2>())>
R operator+(const T &a, const annotated_arg<T2> &b) {
    return a + b.value;
}
#endif

int main() {
    annotated_arg<Custom<int>> annotatedCustom(Custom(42));
    Custom<int> r1 = annotatedCustom + Custom(8);
    Custom<int> r2 = Custom(8) + annotatedCustom;
    return 0;
}
```

(https://godbolt.org/z/TYsfEsEh1)

> R operator/(const annotated_arg<T, PropertyList> &a, const T2 b);

Author: It is strange how this only causes an error when the overloaded operator for `Custom<T>` is a free floating function. I tested this with emulator compilation of the `annotated_arg.cpp` file but I did not see the same issue. I changed the overloaded operators of `MyStruct<T>` to be free floating and tested `MyStruct<T> + annotated_arg<MyStruct<T>>` and it did not produce the invalid operands error. It seems that this problem does not occur with the `annotated_arg` but we may want to add overloaded operators for raw type on lhs to be safe.

template <typename T, typename PropertiesA, typename PropertiesB>
annotated_arg(annotated_arg<T, PropertiesA>, PropertiesB>) ->
    annotated_arg<T, /* a type that combines the properties of PropertiesA and PropertiesB */>;

> // Overloaded operators for raw type

Reviewer: Is there any particular reason why you declared these as free functions but the annotated_arg op annotated_arg versions as member functions?  As far as I can tell the versions that start with annotated_arg won't behave differently regardless of where you declare them because the difference between the two versions is whether implicit conversions are considered but implicit conversions aren't considered anyways for function templates.  Is there something I'm missing or is this decision truly arbitrary?

Author: There does not seem to be any functional differences in this case but it is generally recommended to use non-member function when possible as they offer looser coupling. Additionally, since we the overloaded operators for when the raw type is on the lhs must be non-member functions, using non-member functions for all overloaded operators of annotated_arg and raw type will be more consistent.

Available only if `propertyT` is the property key class of a compile-time
constant property.

> // --- ROW BREAK ---

Reviewer 3: You will also have to add these operators to the section above that has the class skeleton.

a|
[source,c++]
----
template <typename T2, typename PropertyList>

> auto operator+(const annotated_arg<T2, PropertyList> &other) const;

Reviewer 2: Several questions:

* What is the return type for these operators?  Just saying `auto` isn't very helpful.  In particular, is the return type `T` or `annotated_ptr<T>`?  If the latter, what are the properties?
* These operators all assume the RHS is `annotated_arg`.  What if the user wants to do `operator+` where the RHS is a raw type without `annotated_arg`?
* What are the constraints on the type `T2`?
* Can't we just rely on the implicit conversion from `annotated_arg<T>` to `T`, rather than adding all these operators?

Reviewer 3: > Can't we just rely on the implicit conversion from annotated_arg<T> to T, rather than adding all these operators?

Greg, see my comment on the PR (also added to description) regarding this one.

Reviewer 3: > What is the return type for these operators? Just saying auto isn't very helpful.

I think the return type necessarily has to be auto if we want to keep this generic. For example take the case where `annotated_arg` is templated on Intel's `ac_int` library then multiplication returns a wider type the size of which is decided by the `ac_int::operator*`.

> In particular, is the return type T or annotated_ptr<T>? If the latter, what are the properties?

Returned type would not be an `annotated_arg`

Reviewer 2: I think it would help to specify the constraints and the return type better.  For example:

> Available only if `operator+(T2)` is valid for objects of type `T`.  The value and type of the result is the same as the result of `operator+(T2)` applied to the object of type `T`.

Reviewer: > I tested this and it would also require additional overloaded operators with a generic type such as this:

Use SFINAE:
```
template<typename T, typename T2, typename R = decltype(std::declval<T>() + std::declval<T2>())>
R operator+(annotated_arg<T> a, T2 b) {
    return a + b;
};
```
I think this will allow you to use an object of type annotated_arg<T> in any operator+ in which you could have used an object of type T.

Author: @GarveyJoe I tested this notation and the following code works: 
```cpp
template<typename T, typename PropertyList, typename T2, typename R = decltype(std::declval<T>() + std::declval<T2>())>
R operator+(const annotated_arg<T, PropertyList> &a, T2 b) {
    T a1 = a;
    return a1 + b;
};
```

Can you confirm that we want to add these operators in addition to the operators I've currently added for `annotated_arg<T1> + annotated_arg<T2>`?

Reviewer: Yes, confirmed

// --- ROW BREAK ---
a|
[source,c++]
----
template<typename T, typename PropertyList, typename T2, typename R = decltype(std::declval<T>() * std::declval<T2>())>
R operator*(const annotated_arg<T, PropertyList> &a, const T2 b);
----
|
Overloaded `operator*` for annotated_arg and raw type that utilizes the `bperator*` of type `T`.

Available only if `operator*(T2)` is valid for objects of type `T`. The value and type of the result is the same as the result of `operator*(T2)` applied to the object of type `T`.

// --- ROW BREAK ---
a|
[source,c++]
----
template<typename T, typename PropertyList, typename T2, typename R = decltype(std::declval<T>() / std::declval<T2>())>
R operator/(const annotated_arg<T, PropertyList> &a, const T2 b);
----
|
Overloaded `operator/` for annotated_arg and raw type that utilizes the `operator/` of type `T`.

Available only if `operator/(T2)` is valid for objects of type `T`. The value and type of the result is the same as the result of `operator/(T2)` applied to the object of type `T`.

> 

Reviewer 2: Yes, much better!

Available only if `propertyT` is the property key class of a compile-time
constant property.

// --- ROW BREAK ---
a|
[source,c++]
----
template <typename T2, typename PropertyList2,
          typename R = decltype(std::declval<T>() + std::declval<T2>())>
R operator+(const annotated_arg<T2, PropertyList2> &other) const;
----
|
Overloaded `operator+` that utilizes `operator+(T2)` of type `T`.

Available only if `operator+(T2)` is valid for objects of type `T`.
The value and type of the result is the same as the result of `operator+(T2)`

> applied to the object of type `T`.

Reviewer: ```suggestion
applied to the underlying objects of type `T` and `T2`.
```

Available only if `propertyT` is the property key class of a compile-time
constant property.

a|
[source,c++]
----
template <typename T2, typename PropertyList2,
          typename R = decltype(std::declval<T>() OP std::declval<T2>())>
R operatorOP(const annotated_arg<T2, PropertyList2> &other) const;
----
|
Defines the overloaded operators for types `annotated_arg<T, ...>` and
`annotated_arg<T2, ...>`. Let `operatorOP` denote the operator used.

> The overloaded operator `operatorOP` utilizes `operatorOP(T2)` of type `T`
> and is available only if `operatorOP(T2)` is valid for objects of type `T`.
> The value and result is the same as the result of `operatorOP(T2)` applied to the
> objects of type `T` and `T2`.

Reviewer: By my reading of this text, only member functions of T can be called, but that's not what the template checks for and it's not how you implemented it in the header files.  The text implies it was written like this:

```
  template <typename T2, typename PropertyList2,
    typename R = decltype(std::declval<T>() + std::declval<T2>())>
  R operator+(const annotated_arg<T2, PropertyList2> &other) const {
    return obj.operatorOP(other.obj);
  }
```

I think if you want it to match what you implemented you should instead say:

```suggestion
The overloaded operator `operatorOP` utilizes `operatorOP(T, T2)` and is available only if `operatorOP(T, T2)` is well formed.  
The value and result is the same as the result of `operatorOP(T, T2)` applied to the
objects of type `T` and `T2`.
```

|===

> The following table describes the overloaded operators for types `annotated_arg<T, ...>`

Author: @gmlueck I have added some more overloaded operators for logical and bit shift operations. I thought that the spec doc was getting very repetitive so I described the overloaded operators in a generic way before the table. Although, I am wondering whether the table itself is repetitive. We have listed all the overloaded operators in the class definition above so maybe here we can just have the description or have a generic template such as this: 

```c++
template <typename T2, typename PropertyList2,
          typename R = decltype(std::declval<T>() OP std::declval<T2>())>
R operatorOP(const annotated_arg<T2, PropertyList2> &other) const;
```

Reviewer: I like this idea and think it would make the spec more readable.

----

// --- ROW BREAK ---
a|
[source,c++]
----
template <typename T2, typename PropertyList2,
          typename R = decltype(std::declval<T>() & std::declval<T2>())>
R operator&(const annotated_arg<T2, PropertyList2> &other) const;
----

// --- ROW BREAK ---
a|
[source,c++]
----
template <typename T2, typename PropertyList2,
          typename R = decltype(std::declval<T>() \| std::declval<T2>())>
R operator\|(const annotated_arg<T2, PropertyList2> &other) const;
----

// --- ROW BREAK ---
a|
[source,c++]
----
template <typename T2, typename PropertyList2,
          typename R = decltype(std::declval<T>() ^ std::declval<T2>())>
R operator^(const annotated_arg<T2, PropertyList2> &other) const;
----

// --- ROW BREAK ---
a|
[source,c++]
----
template <typename T2, typename PropertyList2,
          typename R = decltype(std::declval<T>() >> std::declval<T2>())>
R operator>>(const annotated_arg<T2, PropertyList2> &other) const;
----

// --- ROW BREAK ---
a|
[source,c++]
----
template <typename T2, typename PropertyList2,
          typename R = decltype(std::declval<T>() << std::declval<T2>())>
R operator<<(const annotated_arg<T2, PropertyList2> &other) const;
----

|===

> The following table describes the overloaded operators for types `annotated_arg<T, ...>`

Author: @gmlueck Also here, I have made the table such that the operator definition with annotated_arg on LHS is on one column and the definition with annotated_arg on RHS is on the other column. Let me know if this looks ok.

|===

> The following template implements the overloaded operators for types

Reviewer: I think the use of the word "template" is confusing here given that it seems you're using it for its English definition when the term "template" has a meaning in C++ and the set of functions you're describing are themselves function templates.

Perhaps just add this as a single row in the table (then it's clear that it refers to a member function) but put operatorOP in italics and in the text in the right side column say something to the effect of "operatorOP can be any of ..." and list all the ops.

`annotated_arg<T, ...>` and `annotated_arg<T2, ...>`. Let `operatorOP` denote
the operator used. The overloaded operator `operatorOP` utilizes `operatorOP(T2)`
of type `T` and is available only if `operatorOP(T2)` is valid for objects of type `T`.

> The value and result is the same as the result of `operator+(T2)` applied to the

Reviewer: ```suggestion
The value and result is the same as the result of `operatorOP(T2)` applied to the
```

objects of type `T` and `T2`.

[source,c++]
----
template <typename T2, typename PropertyList2,
          typename R = decltype(std::declval<T>() OP std::declval<T2>())>
R operatorOP(const annotated_arg<T2, PropertyList2> &other) const;
----

> The following templates implement the overloaded operators for types `annotated_arg<T, ...>`

Reviewer: For the same reason as above, this is a little confusing to read.  It sounds like you're saying that this is the code itself that will be in the header file as opposed to a collapsed representation of it.  Maybe go with:

```suggestion
Free functions of the following form are defined for the `overloaded operators for types `annotated_arg<T, ...>`
```

And, as above, I think you should work in the list of functions somewhere.

and raw type `T2`. Let `operatorOP` denote the operator used. The overloaded operator
`operatorOP` utilizes `operatorOP(T, T2)` and is available only if `operatorOP(T, T2)`
is valid for objects of type `T` and `T2`. The value and result is the same as the result

> of `operator+(T, T2)` applied to the underlying objects of type `T` and `T2`.

Reviewer: ```suggestion
of `operatorOP(T, T2)` applied to the underlying objects of type `T` and `T2`.
```

|===

> Free functions of the following form are defined for the overloaded operators for

Reviewer: If you aren't going to list the overloaded operators I think you at least need to say:
```suggestion
Free functions of the following form are defined for the overloaded operators listed above for
```

types `annotated_arg<T, ...>` and `annotated_arg<T2, ...>`. Let `operatorOP` denote
the operator used. The overloaded operator `operatorOP` utilizes `operatorOP(T2)`
of type `T` and is available only if `operatorOP(T2)` is valid for objects of type `T`.
The value and result is the same as the result of `operatorOP(T2)` applied to the
objects of type `T` and `T2`.

[source,c++]
----
template <typename T2, typename PropertyList2,
          typename R = decltype(std::declval<T>() OP std::declval<T2>())>
R operatorOP(const annotated_arg<T2, PropertyList2> &other) const;

> ----

Reviewer: This still isn't in the table.

Author: Added the table.
